File: src/main/java/cz/wa/secretagent/game/controller/menucreator/MainMenuCreator.java
//package cz.wa.secretagent.game.controller.menucreator;
//
//import java.io.File;
//import java.io.IOException;
//import java.util.ArrayList;
//import java.util.Collection;
//import java.util.Collections;
//import java.util.List;
//
//import org.apache.commons.io.FileUtils;
//import org.slf4j.Logger;
//import org.slf4j.LoggerFactory;
//import org.springframework.beans.factory.annotation.Required;
//
//import cz.wa.secretagent.game.starter.CampaignStarter;
//import cz.wa.secretagent.io.FileSettings;
//import cz.wa.secretagent.io.campaign.CampaignProperties;
//import cz.wa.secretagent.io.campaign.CampaignPropertiesParser;
//import cz.wa.secretagent.launcher.SecretAgentMain;
//import cz.wa.secretagent.menu.builder.TextButtonDescriptor;
//import cz.wa.secretagent.menu.window.GFrame;
//import cz.wa.secretagent.menu.window.component.selectable.GButtonListener;
//
///**
// * Creates main menu. Caches all frames.
// *
// * @author Ondrej Milenovsky
// */
//public class MainMenuCreator extends GeneralMenuCreator {
//
//    private static final long serialVersionUID = 3451448102156807094L;
//    private static final Logger logger = LoggerFactory.getLogger(MainMenuCreator.class);
//
//    private static final String ABOUT_MESSAGE = "About Secret Agent Remake\n\nVersion: "
//            + SecretAgentMain.VERSION + "\nAuthor: Ondrej Milenovsky"
//            + "\nLevel parsing inspired\n    by Camoto Studio\nAdditional images from games:\n    Duke Nukem 3D\n    Commando";
//
//    private FileSettings fileSettings;
//    private CampaignStarter campaignStarter;
//
//    private transient GFrame mainMenu;
//    private transient GFrame aboutDialog;
//
//    public GFrame getMainMenu() {
//        if (mainMenu == null) {
//            mainMenu = createMainMenu();
//            aboutDialog = createAboutDialog();
//        }
//        return mainMenu;
//    }
//
//    private GFrame createMainMenu() {
//        // buttons
//        List<TextButtonDescriptor> buttons = new ArrayList<TextButtonDescriptor>();
//        buttons.add(new TextButtonDescriptor("Load game", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                // TODO load all saved games
//            }
//        }));
//        buttons.add(new TextButtonDescriptor("New game", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                worldHolder.getMenuHolder().addFrame(createNewGameMenu());
//            }
//        }));
//        buttons.add(createSettingsButton());
//        buttons.add(new TextButtonDescriptor("About", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                worldHolder.getMenuHolder().addFrame(aboutDialog);
//            }
//        }));
//        buttons.add(createExitButton());
//
//        // frame
//        GFrame dialog = dialogBuilder.createDialog("Secret Agent Remake", buttons, frameColor, null, false);
//
//        return dialog;
//    }
//
//    private TextButtonDescriptor createSettingsButton() {
//        return new TextButtonDescriptor("Settings", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                GFrame settingsMenu = getSettingsMenu();
//                settingsMenu.setSelectedIndex(0);
//                worldHolder.getMenuHolder().addFrame(settingsMenu);
//            }
//        });
//    }
//
//    private TextButtonDescriptor createExitButton() {
//        return new TextButtonDescriptor("Exit the game", new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                if (gameSettings.confirmDialogs) {
//                    GFrame exitDialog = getExitDialog();
//                    exitDialog.setSelectedIndex(0);
//                    worldHolder.getMenuHolder().addFrame(exitDialog);
//                } else {
//                    exitGame();
//                }
//            }
//        });
//    }
//
//    private GFrame createNewGameMenu() {
//        // files
//
//        Collection<File> files = FileUtils.listFiles(new File(fileSettings.campaignsDir),
//                new String[] { fileSettings.campaignExt }, false);
//
//        // buttons
//        List<TextButtonDescriptor> buttons = new ArrayList<TextButtonDescriptor>(files.size());
//        for (File file : files) {
//            TextButtonDescriptor button = createCampaignButton(file);
//            if (button != null) {
//                buttons.add(button);
//            }
//        }
//
//        // frame
//        GButtonListener closeListener = new GButtonListener() {
//            @Override
//            public void actionPerformed() {
//                worldHolder.getMenuHolder().removeTopFrame();
//            }
//        };
//        GFrame dialog = dialogBuilder.createDialog("Select campaign to start", buttons, frameColor,
//                closeListener, false);
//
//        return dialog;
//    }
//
//    /**
//     * Loads the file and reads campaign name. On exception logs the exception and returns null.
//     * @param file file containing campaign description
//     * @return new button descriptor or null
//     */
//    private TextButtonDescriptor createCampaignButton(final File file) {
//        try {
//            CampaignProperties cp = new CampaignPropertiesParser(file, fileSettings).parse();
//            return new TextButtonDescriptor(cp.getTitle(), new GButtonListener() {
//                @Override
//                public void actionPerformed() {
//                    if (campaignStarter.startCampaign(file)) {
//                        worldHolder.getMenuHolder().removeAllFrames();
//                    }
//                }
//            });
//        } catch (IOException e) {
//            logger.error("Failed to create campaign button", e);
//            return null;
//        }
//    }
//
//    private GFrame createAboutDialog() {
//        return dialogBuilder.createDialog(ABOUT_MESSAGE, Collections.<TextButtonDescriptor> emptyList(),
//                frameColor, new GButtonListener() {
//                    @Override
//                    public void actionPerformed() {
//                        worldHolder.getMenuHolder().removeTopFrame();
//                    }
//                }, false);
//    }
//
//    public FileSettings getFileSettings() {
//        return fileSettings;
//    }
//
//    @Required
//    public void setFileSettings(FileSettings fileSettings) {
//        this.fileSettings = fileSettings;
//    }
//
//    public CampaignStarter getCampaignStarter() {
//        return campaignStarter;
//    }
//
//    @Required
//    public void setCampaignStarter(CampaignStarter campaignStarter) {
//        this.campaignStarter = campaignStarter;
//    }
//
//}

File: src/main/kotlin/secretAgent/game/controller/menuCreator/MainMenuCreator.kt
package secretAgent.game.controller.menuCreator

import org.apache.commons.io.FileUtils
import org.slf4j.LoggerFactory
import secretAgent.game.starter.CampaignStarter
import secretAgent.io.FileSettings
import secretAgent.io.campaign.CampaignPropertiesParser
import secretAgent.launcher.SecretAgentMain
import secretAgent.menu.TextButtonDescriptor
import secretAgent.menu.window.GFrame
import secretAgent.menu.window.component.GButtonListener
import java.io.File
import java.io.IOException
import java.util.ArrayList

/**
 * Creates main menu. Caches all frames.
 *
 * @author Ondrej Milenovsky
 */
class MainMenuCreator : GeneralMenuCreator() {

    lateinit var fileSettings: FileSettings
    lateinit var campaignStarter: CampaignStarter

    @Transient
    private lateinit var mainMenu: GFrame
    @Transient
    private lateinit var aboutDialog: GFrame

    private fun createMainMenu(): GFrame {
        // buttons
        val buttons = arrayListOf(
                TextButtonDescriptor("Load game") {
                    // TODO load all saved games
                },
                TextButtonDescriptor("New game") { worldHolder.menuHolder.addFrame(createNewGameMenu()) },
                createSettingsButton(),
                TextButtonDescriptor("About") { worldHolder.menuHolder.addFrame(aboutDialog) },
                createExitButton())
        // frame

        return dialogBuilder.createDialog("Secret Agent Remake", buttons, frameColor, null, false)
    }

    fun getMainMenu(): GFrame { // TODO lazy
        if (!::mainMenu.isInitialized)
            mainMenu = createMainMenu()
        return mainMenu
    }

    fun getAboutDialog(): GFrame { // TODO lazy
        if (!::aboutDialog.isInitialized)
            aboutDialog = createAboutDialog()
        return aboutDialog
    }

    private fun createSettingsButton() = TextButtonDescriptor("Settings") {
        settingsMenu.selectedIndex = 0
        worldHolder.menuHolder.addFrame(settingsMenu)
    }

    private fun createExitButton() = TextButtonDescriptor("Exit the game") {
        if (gameSettings.confirmDialogs) {
            exitDialog.selectedIndex = 0
            worldHolder.menuHolder.addFrame(exitDialog)
        } else
            exitGame()
    }

    private fun createNewGameMenu(): GFrame {
        // files

        val files = FileUtils.listFiles(File(fileSettings.campaignsDir),
                arrayOf(fileSettings.campaignExt), false)

        // buttons
        val buttons = ArrayList<TextButtonDescriptor>(files.size)
        for (file in files)
            createCampaignButton(file)?.let { buttons += it }

        // frame
        val closeListener: GButtonListener  = { worldHolder.menuHolder.removeTopFrame() }

        return dialogBuilder.createDialog("Select campaign to start", buttons, frameColor, closeListener, false)
    }

    /**
     * Loads the file and reads campaign name. On exception logs the exception and returns null.
     * @param file file containing campaign description
     * @return new button descriptor or null
     */
    private fun createCampaignButton(file: File): TextButtonDescriptor? {
        return try {
            val cp = CampaignPropertiesParser(file, fileSettings).parse()
            TextButtonDescriptor(cp.title) {
                if (campaignStarter.startCampaign(file))
                    worldHolder.menuHolder.removeAllFrames()
            }
        } catch (e: IOException) {
            logger.error("Failed to create campaign button", e)
            null
        }
    }

    private fun createAboutDialog(): GFrame = dialogBuilder.createDialog(ABOUT_MESSAGE, emptyList(),
            frameColor, { worldHolder.menuHolder.removeTopFrame() }, false)

    companion object {

        private const val serialVersionUID = 3451448102156807094L
        private val logger = LoggerFactory.getLogger(MainMenuCreator::class.java)

        private const val ABOUT_MESSAGE = ("""
            About Secret Agent Remake

            Version: ${SecretAgentMain.VERSION}
            Author: Giuseppe Barbieri
            Author original remake: Ondrej Milenovsky
            Level parsing inspired
                by Camoto Studio
            Additional images from games:
                Duke Nukem 3D
                Commando""")
    }
}